mne/inverse_sparse/mxne_optim.py

```diff
@@ -771,237 +894,167 @@ def _tf_mixed_norm_solver_bcd_(M, G, Z, active_set, alpha_space, alpha_time,
                 X_j_new += X_j
 
             rows_norm = linalg.norm(X_j_new, 'fro')
-            if rows_norm <= alpha_space_lc[j]:
+            if rows_norm <= alpha_space_lc[jj]:
                 if was_active:
-                    Z[j] = 0.0
+                    Z[jj] = 0.0
                     active_set_j[:] = False
             else:
                 if was_active:
                     Z_j_new = Z_j + phi(GTR)
                 else:
                     Z_j_new = phi(GTR)
-
                 col_norm = np.sqrt(np.sum(np.abs(Z_j_new) ** 2, axis=0))
 
-                if np.all(col_norm <= alpha_time_lc[j]):
-                    Z[j] = 0.0
+                if np.all(col_norm <= alpha_time_lc[jj]):
+                    Z[jj] = 0.0
                     active_set_j[:] = False
                 else:
                     # l1
-                    shrink = np.maximum(1.0 - alpha_time_lc[j] / np.maximum(
-                                        col_norm, alpha_time_lc[j]), 0.0)
+                    shrink = np.maximum(1.0 - alpha_time_lc[jj] / np.maximum(
+                                        col_norm, alpha_time_lc[jj]), 0.0)
                     Z_j_new *= shrink[np.newaxis, :]
 
                     # l21
                     shape_init = Z_j_new.shape
                     Z_j_new = Z_j_new.reshape(*shape)
                     row_norm = np.sqrt(stft_norm2(Z_j_new).sum())
-                    if row_norm <= alpha_space_lc[j]:
-                        Z[j] = 0.0
+                    if row_norm <= alpha_space_lc[jj]:
+                        Z[jj] = 0.0
                         active_set_j[:] = False
                     else:
-                        shrink = np.maximum(1.0 - alpha_space_lc[j] /
+                        shrink = np.maximum(1.0 - alpha_space_lc[jj] /
                                             np.maximum(row_norm,
-                                            alpha_space_lc[j]), 0.0)
+                                            alpha_space_lc[jj]), 0.0)
                         Z_j_new *= shrink
-                        Z[j] = Z_j_new.reshape(-1, *shape_init[1:]).copy()
+                        Z[jj] = Z_j_new.reshape(-1, *shape_init[1:]).copy()
                         active_set_j[:] = True
-                        R -= np.dot(G_j, phiT(Z[j]))
-
-                        if log_objective:
-                            val_norm_l21_tf += norm_l21_tf(
-                                Z[j], shape, n_orient)
-                            val_norm_l1_tf += norm_l1_tf(
-                                Z[j], shape, n_orient)
+                        R -= np.dot(G_j, phiT(Z[jj]))
+
+        if log_objective:
+            if (ii + 1) % 10 == 0:
+                Zd = np.vstack([Z_ for Z_ in Z.values() if np.any(Z_)])
```

Comment:
[Z[pos] for pos in range(n_positions) if np.any(Z[pos])] is probably a bit slower (neglectable compared to what comes next) but guarantees that the order of positions is correct.